quickwit/quickwit-search/src/leaf.rs

```diff
@@ -1000,134 +995,220 @@ impl CanSplitDoBetter {
     ///
-    /// The leaf search code contains some logic that makes it possible to skip entire splits
-    /// when we are confident they won't make it into top K.
-    /// To make this optimization as potent as possible, we sort the splits so that the first splits
-    /// are the most likely to fill our Top K.
-    /// In the future, as split get more metadata per column, we may be able to do this more than
-    /// just for timestamp and "unsorted" request.
-    fn optimize_split_order(&self, splits: &mut [SplitIdAndFooterOffsets]) {
-        match self {
-            CanSplitDoBetter::SplitIdHigher(_) => {
-                splits.sort_unstable_by(|a, b| b.split_id.cmp(&a.split_id))
+    /// If there's a time range and not enough splits contain at least the number of requested
+    /// documents, return None.
+    fn get_min_required_splits(
+        splits: &[SplitIdAndFooterOffsets],
+        request: &SearchRequest,
+    ) -> Option<usize> {
+        let num_requested_docs = request.start_offset + request.max_hits;
+
+        let mut min_required_splits = 0;
+        let mut partial_sum = 0;
+
+        for split in splits.iter() {
+            if !Self::is_split_contained_in_search_time_range(split, request) {
+                continue;
             }
-            CanSplitDoBetter::SplitTimestampHigher(_)
-            | CanSplitDoBetter::FindTraceIdsAggregation(_) => {
-                splits.sort_unstable_by_key(|split| std::cmp::Reverse(split.timestamp_end()))
+
+            partial_sum += split.num_docs;
+
+            min_required_splits += 1;
+            if partial_sum >= num_requested_docs {
+                return Some(min_required_splits);
             }
-            CanSplitDoBetter::SplitTimestampLower(_) => {
-                splits.sort_unstable_by_key(|split| split.timestamp_start())
+        }
+
+        None
+    }
+
+    fn optimize_split_id_higher(
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
+        splits.sort_unstable_by(|a, b| b.split_id.cmp(&a.split_id));
+
+        if !is_simple_all_query(&request) {
+            // no optimization opportunity here.
+            return Ok(Self::to_splits_with_request(splits, request));
+        }
+
+        let Some(min_required_splits) = Self::get_min_required_splits(&splits, &request) else {
+            // not enough splits contained in time range.
+            return Ok(Self::to_splits_with_request(splits, request));
+        };
+
+        let mut split_with_req = Self::to_splits_with_request(splits, request);
+
+        // In this case there is no sort order, we order by split id.
+        // If the the first split has enough documents, we can convert the other queries to
+        // count only queries.
+        for (_split, request) in split_with_req.iter_mut().skip(min_required_splits) {
```

Comment:
i think this is now wrong due to `get_min_required_splits` skipping over some splits.
Imagine a 2 split scenario, split A contains 2 documents, one in time range and one not, split B contains 10 docs all in range, we want 10 docs. `to_splits_with_request` isn't fully contained in the time range, so it's ignored, B is in range, so it is counted, we get that one split is enough to get the docs we want. Now here B get disabled, because it isn't the 1st split, so only a single doc from A gets returned

Reply:
oh nice catch, that's an easy fix, just `min_required_splits += 1` when `!Self::is_split_contained_in_search_time_range(split, request) {`